### PR TITLE
FlameGraph: Refactor handling of the labels

### PR DESCRIPTION
**Pull request by aocenas**
When we moved to the enum field type in the dataFrame we have to use displayProcessor to handle the values to get the actual value from the enum array. As we still need to support dataFrames where labels aren't in enum fields this created a bit of tedious code where special handler functions needed to be passed everywhere. Instead, this encapsulates all this into a single data wrapper object with appropriate accessors to make sure we handle the labels the same everywhere.

### CODE REVIEW

File: public/app/plugins/panel/flamegraph/components/FlameGraph/FlameGraph.test.tsx [hunk 1]
```diff
@@ -47,7 +47,6 @@ describe('FlameGraph', () => {
         setRangeMin={setRangeMin}
         setRangeMax={setRangeMax}
         selectedView={selectedView}
-        getLabelValue={(val) => val.toString()}
```
Comment (aocenas) [inline public/app/plugins/panel/flamegraph/components/FlameGraph/FlameGraph.test.tsx:50]: One benefit is we don't have to send these getters everywhere separately.

File: public/app/plugins/panel/flamegraph/components/FlameGraph/FlameGraph.tsx [hunk 2]
```diff
@@ -66,52 +65,35 @@ const FlameGraph = ({
...
 }: Props) => {
   const styles = getStyles(selectedView, app, flameGraphHeight);
-  const totalTicks = data.fields[1].values.get(0);
-  const valueField =
-    data.fields.find((f) => f.name === 'value') ?? data.fields.find((f) => f.type === FieldType.number);
-
-  if (!valueField) {
-    throw new Error('Malformed dataFrame: value field of type number is not in the query response');
-  }
+  const totalTicks = data.getValue(0);
```
Comment (aocenas) [inline public/app/plugins/panel/flamegraph/components/FlameGraph/FlameGraph.tsx:70]: Also we don't have to search for the right fields on multiple places

File: public/app/plugins/panel/flamegraph/components/FlameGraph/dataTransform.ts [hunk 3]
```diff
@@ -1,35 +1,126 @@
...
       start: offset,
     };
 
-    levels[item.level].push(newItem);
+    levels[currentLevel].push(newItem);
   }
   return levels;
 }
+
+export class FlameGraphDataContainer {
```
Comment (aocenas) [inline public/app/plugins/panel/flamegraph/components/FlameGraph/dataTransform.ts:45]: Main class to encapsulate the getting the correct values and formatting them.

**Review by adrapereira — APPROVED** [review]
Great job with the refactor! The code was getting messy with all the get functions being passed everywhere so this change is definitely welcome. Left a few minor comments but can approve the PR otherwise.

File: public/app/plugins/panel/flamegraph/components/FlameGraph/dataTransform.ts [hunk 4]
```diff
@@ -1,35 +1,126 @@
...
+export type LevelItem = { start: number; itemIndex: number };
 
 /**
  * Convert data frame with nested set format into array of level. This is mainly done for compatibility with current
  * rendering code.
- * @param dataView
  */
-export function nestedSetToLevels(dataView: DataFrameView<Item>): ItemWithStart[][] {
-  const levels: ItemWithStart[][] = [];
+export function nestedSetToLevels(data: FlameGraphDataContainer): LevelItem[][] {
```
[the remark this answers is not in the document]
Comment (aocenas) [reply public/app/plugins/panel/flamegraph/components/FlameGraph/dataTransform.ts:19]: yes, changed

File: public/app/plugins/panel/flamegraph/components/FlameGraph/dataTransform.ts [hunk 5]
```diff
@@ -1,35 +1,126 @@
...
+  const levels: LevelItem[][] = [];
   let offset = 0;
 
-  for (let i = 0; i < dataView.length; i++) {
+  for (let i = 0; i < data.data.length; i++) {
     // We have to clone the items as .get(i) returns a changing pointer not the data themselves.
```
[the remark this answers is not in the document]
Comment (aocenas) [reply public/app/plugins/panel/flamegraph/components/FlameGraph/dataTransform.ts:24]: Removed

**Pull request merged**